deployLOCAL/runETL.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Apr  9 22:45:43 2019
+
+@author: zackkingbackup
+"""
+import sys
+sys.path.append('/Users/zackkingbackup/Documents/Job Search/CapitalOne/sample-project-data-engineer/src/')
+from Extract import extract_function
+from Transform import transform_function
+from Load import load_function
+import argparse
+import boto3
+import os
+
+''' use argparse.parser to pick off command line args '''
+parser = argparse.ArgumentParser(prog='runETL',
+                                 description='Run Sample Project')
+
+parser.add_argument('-d', dest='date',
+                    default='1999-12-31',help='Date in YYYY-MM-DD format'+
+                    '; default=1999-12-31',
+                    metavar='YYYY-MM-DD')
+parser.add_argument('-t', dest='local_temp',
+                    default='temp_data',
+                    help='relative path for directory to temporarily store '+
+                    'files downloaded '+
+                    'from the S3 bucket.\nOnly needed if running locally;\n'+
+                    'default=temp_data',
+                    metavar='dir')
+parser.add_argument('-C', dest='clean',
+                    default='False',help='Boolean to determine if local temp '+
+                    'space to store S3 downloads\nshould be deleted on script'+
+                    ' completion;\ndefault=False (do not remove it)',
+                    metavar='True/False')
+
+args = parser.parse_args()
+
+QUERY_DATE = args.date
+local_temp = args.local_temp
+clean = args.clean
+
+''' the temp data space needs to end with a forward slash '''
+if local_temp[-1] != '/':
+    local_temp += '/'
+
+LOCAL = True
+
+if os.path.exists(os.path.join(os.getcwd(), local_temp)):
+    ''' see if the local temp space already has files in it; delete them '''
+    for temp_file in os.listdir(local_temp):
+        file_path = os.path.join(local_temp, temp_file)
+        if os.path.isfile(file_path):
+            os.unlink(file_path)
+else:
+    ''' if the local temp space does not exist, create it '''
+    os.mkdir(os.path.join(os.getcwd(), local_temp))
+    
+if clean != 'False':
+    CLEAN = True
+else:
+    CLEAN = False
+    
+''' set up an S3 resource to use for file download/upload to S3 bucket '''
+s3_resource = boto3.resource('s3')
+
+''' execute the ETL functions '''
+extract_function(QUERY_DATE,s3_resource,LOCAL=LOCAL,
```

Comment:
I wonder whether code for S3 download/uploads should be in this script, rather than in the Extract, Transform, Load modules.  I tend to think that those modules' contracts should be in terms of either Python objects or local files.  That would make them easier to test, for sure.

So for example, the `extract` function could return a DataFrame, as would `transform`.  This is somewhat overkill for the extract requirements for this very simple sample project, but in a real ETL, I would want the extract code to be testable separately from the code that reads/writes to S3.

Reply:
Agree 100%. I think I misinterpreted the README by assuming the modules had to execute their own S3 download/uploads. Something about creating a new S3 resource instance in each module didn't sit quite right with me, which led to me passing the S3 resource in as an argument despite the added annoyance in testing. I should have asked for clarification earlier.

Reply:
Despite our concurrence that in a real ETL, the uploads/downloads should be done in the end-to-end script, do you mind if I leave this part of my code as is? Your comment that this is overkill for what I'm trying to do makes me think 'yes'.

Reply:
@zking1219 yeah I think leave this part of the code as-is.